macaroonbakery/bakery.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2016 Canonical Ltd.
+# Licensed under the AGPLv3, see LICENCE file for details.
+
+import base64
+import json
+import requests
+import webbrowser
+
+import macaroons
+
+
+def deserialize(json_macaroon):
+    '''Deserialize a JSON macaroon into a macaroon object.
+
+    @param the JSON macaroon to deserialize.
+    @return the deserialized macaroon.
+    '''
+    return macaroons.deserialize(to_base64(json_macaroon))
+
+
+def to_base64(json_macaroon):
+    '''Base 64 encode a json macaroon
+
+    @param the JSON macaroon to encode.
+    @return the encoded macaroon.
+    '''
+    return base64.urlsafe_b64encode(json.dumps(json_macaroon))
+
+
+def add_padding(s):
+    '''Add padding to base64 encoded string
+    libmacaroon does not give padded base64 string from serialization
+
+    @param string s to be padded
+    @return a padded string
+    '''
+    return s + '=' * ((4 - (len(s) % 4)) % 4)
+
+
+def serialize_macaroon_string(macaroon):
+    '''Serialize macaroon to string
+
+    @param macaroon to be serialized
+    @return a string serialization form of the macaroon
+    '''
+    return base64.urlsafe_b64decode(
+        add_padding(macaroon.serialize_json()))
+
+
+class DischargeError(Exception):
+    pass
+
+
+class Bakery(object):
+    ''' Bakery is a wrapper around requests that ensures
+    that macaroons headers are sent to the server if present or gather the
+    discharged macaroons automatically before sending the request.
+    '''
+    def __init__(self, initial_macaroons=None, discharge_token=None,
+                 timeout=30):
+        '''
+        @param initial_macaroons initial macaroon to be used in headers.
+        @param discharge_token initial discharge token.
+        @param timeout timeout for wait url.
+        '''
+        self.macaroons = initial_macaroons
+        self.discharge_token = discharge_token
+        self.timeout = timeout
+
+    def get(self, url, **kwargs):
+        '''Act as the same than requests.get, same arguments,
+        same returned value, same exception raised except DischargeError
+        when something wrong happened when discharging.
+        If macaroons are present, the headers will be augmented with it.
+        '''
+        if self.macaroons is not None:
+            if 'headers' not in kwargs:
+                kwargs['headers'] = {}
+            kwargs['headers']['Bakery-Protocol-Version'] = 1
+            kwargs['headers']['Macaroons'] = self.macaroons
+
+        response = requests.get(url, **kwargs)
+        if (response.status_code == 401 and
+                response.headers['Www-Authenticate'] == 'Macaroon') \
+                or response.status_code == 407:
+            try:
+                response_json = response.json()
+            except ValueError:
+                raise DischargeError(
+                    'Unable to get data from unauthorized response')
+            if 'Code' not in response_json:
```

Comment:
I'd keep using try/except blocks here, as we could assume the response to be valid 99% of the times, and in this case it's better to ask forgiveness than permission.
For instance:
```python
try:
    code = response_json['Code']
except KeyError:
    raise DischargeError(
        'code not found in unauthorized response {!r}'.format(response.content))
if code != 'macaroon discharge required':
    # What happens in this case? It's not handled in current code.
    return response
try:
    serialized_macaroon = response_json['Info']['Macaroon']
except KeyError:
    raise DischargeError(
        'macaroon not found in unauthorized response {!r}'.format(response.content))
macaroon = deserialize(serialized_macaroon)
...  # Etc...
```